### PR TITLE
JDK-8313764: Offer JVM HS functionality to shared lib load operations done by the JDK codebase

### PR DESCRIPTION
**Pull request by MBaesken**
Currently there is a number of functionality that would be interesting to have for shared lib load operations in the JDK C code.
Some examples :
Events::log_dll_message for hs-err files reporting
JFR event NativeLibraryLoad
There is the need to update the shared lib Cache on AIX ( see LoadedLibraries::reload() , see also https://bugs.openjdk.org/browse/JDK-8314152 ),
this is currently not fully in sync with libs loaded form jdk c-libs and sometimes reports outdated information

Offer an interface (e.g. jvm.cpp) to support this.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313764](https://bugs.openjdk.org/browse/JDK-8313764): Offer JVM HS functionality to shared lib load operations done by the JDK codebase (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15264/head:pull/15264` \
`$ git checkout pull/15264`

Update a local copy of the PR: \
`$ git checkout pull/15264` \
`$ git pull https://git.openjdk.org/jdk.git pull/15264/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15264`

View PR using the GUI difftool: \
`$ git pr show -t 15264`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15264.diff">https://git.openjdk.org/jdk/pull/15264.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15264#issuecomment-1676881491)

### CODE REVIEW

**Comment by AlanBateman** [comment]
Having dlopen usages in non-core native libraries such as such as libsctp, libawt_xawt, ... have their dlopen usages compiled to use dlopen_ext in libjvm introduces questionable coupling that I think requires broader discussion. For the JFR NativeLibraryLoad event then I suppose there is a discussion on whether events for statically or dynamically loaded libs is interesting or not as there isn't a corresponding System.loadLibrary or SymbolLookup.libraryLookup. JDK-8314152 seems to very AIX specific with shared lib caching that I would hope complicate interfaces for other ports.

**Comment by prrace** [comment]
I don't see why we need this .. 
The desktop module has a LOT of calls to dlopen() .. I wasn't sure why you were only interested in two of them.
I suspect because there's #ifdef AIX blocks in both of these files that do AIX-specific things for the dlopen .. but I'm still not sure I understand .. why is what you are doing only wanted for the AIX cases ? 

And I'd prefer that if anything is done that whatever you need to do be entirely within that #ifdef AIX block and so not affect anything else.

**Comment by MBaesken** [comment]
Hi Phil, yes there are of course more call sites of dlopen in the JDK C codebase.
I could cover them too, if this is wished.

> . but I'm still not sure I understand .. why is what you are doing only wanted for the AIX cases ?

The lib cache update ` LoadedLibraries::reload();`   is  AIX specific;  the UL and JFR events are cross platform.

( Btw. in our proprietary JVM we have this approach for many years )

**Comment by TheRealMDoerr** [comment]
Please check windows-aarch64 build error: unresolved external symbol dlopen_ext

**Comment by MBaesken** [comment]
> Please check windows-aarch64 build error: unresolved external symbol dlopen_ext

Hi Martin, thanks !  I did a small adjustment, now the windows aarch64 build works.

**Comment by AlanBateman** [comment]
> yes this is of course AIX specific. However I found something that might be similar on Windows, there we have in case of successful LoadLibrary in os::dll_load calls to SymbolEngine::recalc_search_path(); However I did not check the details on this; but for AIX we have a real problem with outdated/incomplete stacks in hs_err files.

Right now, you've got concerns/objections from two potential reviewers so it would be better to see if there are other options to help the error logs on AIX.

**Comment by dholmes-ora** [comment]
Sorry but looking at the hotspot part of this I do not like the code in jvm.cpp at all - it is  far too messy. I expected to see a simple interface to os::dlopen which then handles all the platform specific issues.

I'm also somewhat dubious about having all the JDK code use a VM interface for this, The usual coupling to the VM is to just provide native implementations of core library methods, not to have the VM provide a general purpose utility interface like dynamic library loading.

**Comment by MBaesken** [comment]
> Sorry but looking at the hotspot part of this I do not like the code in jvm.cpp at all - it is far too messy. I expected to see a simple interface to os::dlopen which then handles all the platform specific issues.
> 
> I'm also somewhat dubious about having all the JDK code use a VM interface for this, The usual coupling to the VM is to just provide native implementations of core library methods, not to have the VM provide a general purpose utility interface like dynamic library loading.


Hi David, so would you prefer to call into 'os::dll_load' (probably you meant that) instead ?  That would indeed make the coding smaller and less 'messy' .
However 'os::dll_load' has no flags parameter but uses a default, have to check if this might be a problem ...

**Comment by MBaesken** [comment]
Hello, any comments about the idea of calling into 'os::dll_load' instead ? That would indeed make the coding smaller and less 'messy' .

**Comment by dholmes-ora** [comment]
@MBaesken I'm not at all sure what it would look like - sorry. But there doesn't seem to be any general support from the library folk for this.

**Comment by MBaesken** [comment]
In the meantime,   https://bugs.openjdk.org/browse/JDK-8295159  was integrated.
There we restore on x86_64 and aarch the floating point environment  (fenv_t)  after loading libs in HS os::dll_load / dlopen_helper to avoid  'silent' manipulation of the fenv by 'bad' shared libs.
But unfortunately this does not work for  libs  not going through this coding (like the ones loaded from JDK c code).
Might be worth considering in context of this change .

File: src/java.desktop/unix/native/common/awt/fontpath.c [hunk 1]
```diff
@@ -48,6 +48,9 @@
 #define AWT_UNLOCK()
 #endif /* !HEADLESS */
 
+/* for dlopen */
+#include <macro_helper.h>
+
```
Comment (prrace) [inline src/java.desktop/unix/native/common/awt/fontpath.c:53]: I have no idea why we would want to use some VM internal in the desktop module.
Please take the desktop module changes out of this PR.

**Pull request closed without merging**